# File: 06_samtools_array_job.R
# Auth: umar.niazi@kcl.as.uk
# DESC: create a parameter file and shell script to run array job on hpc
# Date: 24/9/2021


## set variables and source libraries
source('header.R')

dfData = read.csv(file.choose(), header=T)

# set header variables 
cvShell = '#!/bin/bash -l'
cvJobName = '#SBATCH --job-name=samtools-array'
cvNodes = '#SBATCH --nodes=1'
cvProcessors = '#SBATCH --ntasks=1'
# format d-hh:mm:ss
cvRuntime = '#SBATCH --time=1-05:05:00'
cvPartition = '#SBATCH --partition brc'
# How much memory you need.
# --mem will define memory per node and
# --mem-per-cpu will define memory per CPU/core. Choose one of those.
cvMemoryReserve = '#SBATCH --mem-per-cpu=2000MB'
# Turn on mail notification. There are many possible self-explaining values:
# NONE, BEGIN, END, FAIL, ALL (including all aforementioned)
# For more values, check "man sbatch"
cvMail = '#SBATCH --mail-type=END,FAIL'
# set array job loop
dim(dfData)
cvArrayJob = '#SBATCH --array=1-2'

# set the directory names
cvInput = 'input/'
#cvSam = 'samtools'

# create a parameter file and shell script
dir.create('AutoScripts')
oFile.param = file('AutoScripts/samtools_param.txt', 'wt')

temp = lapply(1:nrow(dfData), function(x){
  lf = list(paste0(as.character(dfData$Run[x]), '_1.fastq'))
  # write samtools command variables
  in.s1 = paste0(cvInput, lf[[1]], '.sam')
  # output file in bam format
  s2 = paste0(cvInput, lf[[1]], '.bam')
  # remove low quality reads below 25
  s3 = paste0(cvInput, lf[[1]], '_q25.bam')
  # sort the file
  s4 = paste0(cvInput, lf[[1]], '_q25_sort.bam')
  # remove duplicates
  s5 = paste0(cvInput, lf[[1]], '_q25_sort_rd.bam')
  p1 = paste(in.s1, s2, s3, s4, s5, sep=' ')
  writeLines(p1, oFile.param)
  # return(data.frame(idSample=dfFiles$idSample[1], name=c(s2, s4, s5), type=c('original bam', 'quality 10 sorted bam',
  #                                                                            'quality 10 sorted bam duplicates removed'), 
  #                   group1=dfFiles$group1[1]))
})

close(oFile.param)

oFile = file('AutoScripts/samtools.sh', 'wt')

writeLines(c(cvShell, cvJobName, cvArrayJob, cvNodes, cvProcessors, 
             cvRuntime, cvPartition, cvMemoryReserve, cvMail), oFile)

writeLines(c('# Autogenerated script from write_samtools_script.R', paste('# date', date())), oFile)
writeLines(c('# make sure directory paths exist before running script'), oFile)
writeLines('\n\n', oFile)

# module load
writeLines(c('module load apps/samtools/1.9.0-singularity'), oFile)
writeLines('\n\n', oFile)

## write array job lines
writeLines("# Parse parameter file to get variables.
           number=$SLURM_ARRAY_TASK_ID
           paramfile=samtools_param.txt
           
           ins1=`sed -n ${number}p $paramfile | awk '{print $1}'`
           bamfile=`sed -n ${number}p $paramfile | awk '{print $2}'`
           bamq25=`sed -n ${number}p $paramfile | awk '{print $3}'`
           bamq25sort=`sed -n ${number}p $paramfile | awk '{print $4}'`
           bamrd=`sed -n ${number}p $paramfile | awk '{print $5}'`
           
           # 9. Run the program.", oFile)

# sam to bam
p1 = paste('samtools view -b -S', '$ins1', '>', '$bamfile', sep=' ')
com1 = paste(p1)
# remove low quality reads
p1 = paste('samtools view -b -q 25', '$bamfile', '>', '$bamq25', sep=' ')
com2 = paste(p1)
# sort the file
p1 = paste('samtools sort -o', '$bamq25sort', '$bamq25', sep=' ')
com3 = paste(p1)
# create index before duplicate removal
p1 = paste('samtools index', '$bamq25sort', sep=' ')
com3.1 = paste(p1)
# remove duplicates, for paired end reads
p1 = paste('samtools rmdup', '$bamq25sort', '$bamrd', sep=' ')
com4 = paste(p1)
# create index
p1 = paste('samtools index', '$bamrd', sep=' ')
com5 = paste(p1)

writeLines(c(com1, com2, com3, com3.1, com4, com5), oFile)
writeLines('\n\n', oFile)

close(oFile)
